File: src/demo/word_distance_demo.cpp
// Copyright (c) 2017, Baidu.com, Inc. All Rights Reserved
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "familia/semantic_matching.h"

#include <gflags/gflags.h>
#include <iostream>
#include <iomanip>

using std::string;
using std::vector;
using std::cin;
using std::cout;
using std::endl;

DEFINE_string(model_dir, "./", "model directory");
DEFINE_string(emb_file, "./", "Topical Word Embedding (TWE) file");
DEFINE_int32(top_k, 20, "the nearest k words");

namespace familia {
// 寻找距离词最邻近的K个词Demo类
class WordDistanceDemo {
public:
    WordDistanceDemo() : _twe(FLAGS_model_dir, FLAGS_emb_file) {
    }

    ~WordDistanceDemo() = default;

    // 获取词距离最近的k个词，并打印出来
    void find_nearest_words(const string& word, int k) {
        vector<WordAndDis> items(k);
        // 如果词不存在模型词典，则返回
        if (!_twe.contains_word(word)) {
            cout << word << " is out of vocabulary." << endl;
            return;
        }

        _twe.nearest_words(word, items);
        print_result(items);
    }

    // 打印结果
    void print_result(const std::vector<WordAndDis>& items) {
        cout << "Word                       Cosine Distance  " << endl;
        cout << "--------------------------------------------" << endl;
        for (const auto& item : items) {
            cout << std::left << std::setw(24) << item.word << "\t" 
                 << item.distance << endl;
        }
        cout << endl;
    }

private:
    // Topic Word Embedding模型
    TopicalWordEmbedding _twe;
};
} // namespace familia

int main(int argc, char* argv[]) {
    GOOGLE_PROTOBUF_VERIFY_VERSION;
    google::SetVersionString("1.0.0.0");
    string usage = string("Usage: ./word_distance_demo --model_dir=\"PATH/TO/MODEL\" ") +
                   string("--emb_file=\"webpage_twe_lda.model\" ") +
                   string("--top_k=\"20\" ");
    google::SetUsageMessage(usage);
    google::ParseCommandLineFlags(&argc, &argv, true);

    familia::WordDistanceDemo wd_demo;

    string word;
    while (true) {
        cout << "请输入词语:\t";
        getline(cin, word);
        wd_demo.find_nearest_words(word, FLAGS_top_k);
    }

    return 0;
}

File: python/demo/word_distance_demo.py
#coding=utf-8

# Copyright (c) 2017, Baidu.com, Inc. All Rights Reserved
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

import sys
from familia_wrapper import TopicalWordEmbeddingsWrapper

if sys.version_info < (3,0):
    input = raw_input

def print_result(result_list):
    # 分隔符
    print("Word\t\t\tCosine Distance")
    print("-" * 40)
    # 判断list是否为空，为空则返回
    if result_list is None:
        return
    for (word, distance) in result_list:
        print("{}\t\t\t{}".format(word, distance))

if __name__ == "__main__":
    if len(sys.argv) < 3:
        sys.stderr.write("Usage: python {} {} {}\n".format(
            sys.argv[0], "model_dir", "emb_file"))
        exit(-1)
    model_dir = sys.argv[1]
    emb_file = sys.argv[2]
    # 创建topical_word_embeddings_wrapper对象
    twe_wrapper = TopicalWordEmbeddingsWrapper(model_dir, emb_file)
    while True:
        word = input("Enter word: ")
        result_list = twe_wrapper.nearest_words(word.strip(), 10)
        print_result(result_list)
